File: app/src/main/java/com/example/musicplayer/MainActivity.java
package com.example.musicplayer;


import android.Manifest;
import android.annotation.SuppressLint;
import android.content.pm.PackageManager;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.media.PlaybackParams;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.RequiresApi;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.graphics.Palette;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.Gravity;
import android.view.View;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.SeekBar;
import android.widget.TextView;
import android.widget.Toast;

import com.sothree.slidinguppanel.SlidingUpPanelLayout;
import com.sothree.slidinguppanel.SlidingUpPanelLayout.PanelSlideListener;
import com.sothree.slidinguppanel.SlidingUpPanelLayout.PanelState;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;


public class MainActivity extends AppCompatActivity implements View.OnClickListener,
        MediaPlayer.OnCompletionListener, SeekBar.OnSeekBarChangeListener {

    public static final int REQUEST_CODE_PERMISSION = 1001;

    private String[] permissionList = { Manifest.permission.WRITE_EXTERNAL_STORAGE,
            Manifest.permission.READ_EXTERNAL_STORAGE };

    private RecyclerView mRecyclerViewSongs;
    private ArrayList<Song> mArrSongs;
    private SlidingUpPanelLayout mSlidingUpLayout;
    private RelativeLayout mRelative;
    private MediaPlayer mMediaPlayer;
    private Handler mHandler;

    //On slidingLayout
    private ImageView mImgPlayingSong;
    private TextView mTxtPlayingSongName;
    private ImageButton mImgBtnPlayOnSlideLay;

    //In SlidingLayout
    private ImageView mImgCurrentPlaySong;
    private ImageButton mImgBtnShuffle, mImgBtnPlay, mImgBtnNext, mImgBtnPrevious, mImgBtnRepeat;
    private TextView mTxtSongPlayCurrDuration, mTxtSongPlayTime;
    private SeekBar mSeekBarPlaySong;

    private int currentSongIndex = 0;
    private boolean isShuffle = false;
    private boolean isRepeat = false;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        requestAppPermissions();

        setContentView(R.layout.activity_main);

        if (!checkPermissionGrantedOrNot()){
            requestAppPermissions();
        }

//        setVolumeControlStream(AudioManager.STREAM_MUSIC);

        Toolbar mToolBar = findViewById(R.id.toolbar);
        setSupportActionBar(mToolBar);

        initRecyclerView();

        getId();
        setListener();

        mHandler = new Handler();
        mMediaPlayer = new MediaPlayer();

        mSeekBarPlaySong.setOnSeekBarChangeListener(this);
        mMediaPlayer.setOnCompletionListener(this);

        mArrSongs = new ArrayList<>();

        final Runnable uiRunnable = new Runnable() {
            @Override
            public void run() {
                AdapterSongs mAdapterSongs = new AdapterSongs(mArrSongs, new AdapterSongs.OnSongClickListener() {
                    @Override
                    public void onSongClick(int position) {
                        currentSongIndex = position;
                        playSong(position);
                    }
                });
                mRecyclerViewSongs.setAdapter(mAdapterSongs);
            }
        };

        Thread thread = new Thread(new Runnable() {
            @Override
            public void run() {
                mArrSongs = SongManager.getMp3Songs(MainActivity.this);
                runOnUiThread(uiRunnable);
            }
        });
        thread.start();

        if (!mArrSongs.isEmpty()) {

            mSlidingUpLayout.addPanelSlideListener(new PanelSlideListener() {

                @Override
                public void onPanelSlide(View panel, float slideOffset) {

                    if (slideOffset > 0.88) {
                        mRelative.setBackgroundColor(getResources().getColor(R.color.whiteTransparent));

                        mImgBtnPlayOnSlideLay.setVisibility(View.INVISIBLE);
                        changeMusicAlbumArt(currentSongIndex);
                        mImgCurrentPlaySong.setScaleType(ImageView.ScaleType.CENTER_CROP);
                    }
                    if (slideOffset < 0.09) {
                        mImgBtnPlayOnSlideLay.setVisibility(View.VISIBLE);
                        mImgCurrentPlaySong.setImageResource(R.drawable.music);
                        getWindow().setStatusBarColor(getResources().getColor(R.color.colorPrimaryDark));
                    }
                }

                @Override
                public void onPanelStateChanged(View panel, PanelState previousState, PanelState newState) {

                }
            });
        }
    }

    private boolean checkPermissionGrantedOrNot() {
        for (String s : permissionList) {
            int res = checkCallingOrSelfPermission(s);
            return (res != PackageManager.PERMISSION_GRANTED);
        }
        return false;
    }

    private void requestAppPermissions() {
        //check which permission granted
        List<String> listPermissionNeeded = new ArrayList<>();
        for (String permission: permissionList){
            if (ContextCompat.checkSelfPermission(MainActivity.this,permission) != PackageManager.PERMISSION_GRANTED){
                listPermissionNeeded.add(permission);
            }
        }

        //ask for non-permission granted
        if (!listPermissionNeeded.isEmpty()){
            ActivityCompat.requestPermissions(MainActivity.this,
                    listPermissionNeeded.toArray(new String[listPermissionNeeded.size()]),REQUEST_CODE_PERMISSION);
            return;
        }
    }

    private void initRecyclerView() {
        mRecyclerViewSongs = findViewById(R.id.recyclerViewSongList);
        mRecyclerViewSongs.setLayoutManager(new LinearLayoutManager(
                this, LinearLayoutManager.VERTICAL, false));
    }

    private void getId() {
        try {
            mRelative = findViewById(R.id.relative);
            mSlidingUpLayout = findViewById(R.id.sliding_layout);

            //on SlideUpPanel Layout
            mImgPlayingSong = findViewById(R.id.img_playing_song);
            mTxtPlayingSongName = findViewById(R.id.txt_playing_songName);
            mImgBtnPlayOnSlideLay = findViewById(R.id.imgBtn_play);

            //In slidePanel Layout
            mImgCurrentPlaySong = findViewById(R.id.img_play_song);
            mImgBtnShuffle = findViewById(R.id.btn_song_shuffle_play);
            mImgBtnRepeat = findViewById(R.id.btn_song_repeat_play);
            mImgBtnPlay = findViewById(R.id.btn_song_play);
            mImgBtnNext = findViewById(R.id.btn_song_next);
            mImgBtnPrevious = findViewById(R.id.btn_song_previous);
            mSeekBarPlaySong = findViewById(R.id.seekBar_play_song);
            mTxtSongPlayTime = findViewById(R.id.txt_song_TotalTime);
            mTxtSongPlayCurrDuration = findViewById(R.id.txt_song_Curr_Time);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void setListener() {
        try {
            mImgBtnPlayOnSlideLay.setOnClickListener(this);
            mImgBtnPlay.setOnClickListener(this);
            mImgBtnNext.setOnClickListener(this);
            mImgBtnPrevious.setOnClickListener(this);
            mImgBtnShuffle.setOnClickListener(this);
            mImgBtnRepeat.setOnClickListener(this);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onClick(View v) {

        switch (v.getId()) {

            case R.id.imgBtn_play:
                try {
                    if (mMediaPlayer.isPlaying()) {
                        if (mMediaPlayer != null) {
                            mMediaPlayer.pause();
                            mImgBtnPlayOnSlideLay.setImageResource(R.drawable.ic_action_play);
                            mImgBtnPlay.setImageResource(R.drawable.ic_action_play);
                        }
                    } else {
                        if (mMediaPlayer != null) {
                            mMediaPlayer.start();
                            mImgBtnPlayOnSlideLay.setImageResource(R.drawable.ic_action_pause);
                            mImgBtnPlay.setImageResource(R.drawable.ic_action_pause);
                        }
                    }
                } catch (IllegalStateException iStateException) {
                    iStateException.printStackTrace();
                }
                break;

            case R.id.btn_song_play:
                try {
                    if (mMediaPlayer.isPlaying()) {

                        if (mMediaPlayer != null) {
                            mMediaPlayer.pause();
                            mImgBtnPlay.setImageResource(R.drawable.ic_action_play);
                            mImgBtnPlayOnSlideLay.setImageResource(R.drawable.ic_action_play);
                        }
                    } else {
                        if (mMediaPlayer != null) {
                            mMediaPlayer.start();
                            mImgBtnPlay.setImageResource(R.drawable.ic_action_pause);
                            mImgBtnPlayOnSlideLay.setImageResource(R.drawable.ic_action_pause);
                        }
                    }
                } catch (IllegalStateException iStateException) {
                    iStateException.printStackTrace();
                }
                break;

            case R.id.btn_song_next:

                if (currentSongIndex < (mArrSongs.size() - 1)) {
                    playSong(currentSongIndex + 1);
                    changeMusicAlbumArt(currentSongIndex + 1);
                    currentSongIndex = currentSongIndex + 1;

                } else {
                    playSong(currentSongIndex);
                }
                break;

            case R.id.btn_song_previous:

                if (currentSongIndex > 0) {
                    playSong(currentSongIndex - 1);
                    changeMusicAlbumArt(currentSongIndex - 1);
                    currentSongIndex = currentSongIndex - 1;

                } else {
                    // play last song
                    playSong(mArrSongs.size() - 1);
                    currentSongIndex = mArrSongs.size() - 1;
                }
                break;
            case R.id.btn_song_shuffle_play:

                if (isShuffle) {
                    isShuffle = false;
                    Toast.makeText(getApplicationContext(), "Shuffle is OFF", Toast.LENGTH_SHORT).show();
                    mImgBtnShuffle.setImageResource(R.drawable.ic_action_shuffle_off);
                } else {
                    // make repeat to true
                    isShuffle = true;
                    Toast.makeText(getApplicationContext(), "Shuffle is ON", Toast.LENGTH_SHORT).show();
                    // make shuffle to false
                    isRepeat = false;
                    mImgBtnShuffle.setImageResource(R.drawable.ic_action_shuffle_on);
                    mImgBtnRepeat.setImageResource(R.drawable.ic_action_repeat_off);
                }
                break;

            case R.id.btn_song_repeat_play:

                if (isRepeat) {
                    isRepeat = false;
                    Toast.makeText(getApplicationContext(), "Repeat is OFF", Toast.LENGTH_SHORT).show();
                    mImgBtnRepeat.setImageResource(R.drawable.ic_action_repeat_off);
                } else {
                    // make repeat to true
                    isRepeat = true;
                    Toast.makeText(getApplicationContext(), "Repeat is ON", Toast.LENGTH_SHORT).show();
                    // make shuffle to false
                    isShuffle = false;
                    mImgBtnRepeat.setImageResource(R.drawable.ic_action_repeat_on);
                    mImgBtnShuffle.setImageResource(R.drawable.ic_action_shuffle_off);
                }
                break;
        }
    }

    /**
     * Update timer on seekbar
     */
    public void updateProgressBar() {
        mHandler.postDelayed(mUpdateTimeTask, 100);
    }

    private Runnable mUpdateTimeTask = new Runnable() {
        @SuppressLint("SetTextI18n")
        @Override
        public void run() {
            long totalDuration = mMediaPlayer.getDuration();
            long currentDuration = mMediaPlayer.getCurrentPosition();

            // Displaying Total Duration time
            mTxtSongPlayTime.setText("" + Utilities.milliSecondsToTimer(totalDuration));
            // Displaying time completed playing
            mTxtSongPlayCurrDuration.setText("" + Utilities.milliSecondsToTimer(currentDuration));

            // Updating progress bar
            int progress = (Utilities.getProgressPercentage(currentDuration, totalDuration));
            //Log.d("Progress", ""+progress);
            mSeekBarPlaySong.setProgress(progress);

            // Running this thread after 100 milliseconds
            mHandler.postDelayed(this, 100);
        }
    };

    @Override
    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {

    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {
        mHandler.removeCallbacks(mUpdateTimeTask);
    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {

        mHandler.removeCallbacks(mUpdateTimeTask);
        int totalDuration = mMediaPlayer.getDuration();
        int currentPosition = Utilities.progressToTimer(seekBar.getProgress(), totalDuration);

        // forward or backward to certain seconds
        mMediaPlayer.seekTo(currentPosition);

        // update timer progress again
        updateProgressBar();
    }

    public void playSong(final int songIndex) {
        try {
            if (mMediaPlayer.isPlaying()) {
                mMediaPlayer.stop();
            }
            mMediaPlayer.reset();

            mMediaPlayer.setDataSource(mArrSongs.get(songIndex).getPath());
//            mMediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
            mMediaPlayer.prepare();
            mMediaPlayer.setOnPreparedListener(new MediaPlayer.OnPreparedListener() {
                @RequiresApi(api = Build.VERSION_CODES.M)
                @Override
                public void onPrepared(MediaPlayer mp) {
                    mp.start();

                    mMediaPlayer.setPlaybackParams(new PlaybackParams().allowDefaults().setAudioFallbackMode(PlaybackParams.AUDIO_FALLBACK_MODE_DEFAULT));
                    mImgPlayingSong.setImageBitmap(mArrSongs.get(songIndex).getAlbumArt());
                    mTxtPlayingSongName.setText(mArrSongs.get(songIndex).getSongTitle());
                    mImgBtnPlayOnSlideLay.setImageResource(R.drawable.ic_action_pause);
                    mImgBtnPlay.setImageResource(R.drawable.ic_action_pause);

                    mSeekBarPlaySong.setProgress(0);
                    mSeekBarPlaySong.setMax(100);

                    updateProgressBar();
                }
            });
            mMediaPlayer.prepareAsync();

        } catch (IOException e) {
            e.printStackTrace();
        } catch (IllegalStateException iState) {
            iState.printStackTrace();
        }
    }

    @Override
    public void onCompletion(MediaPlayer mp) {
        // check for repeat is ON or OFF
        if (isRepeat) {
            // repeat is on play same song again
            playSong(currentSongIndex);

        } else if (isShuffle) {
            // shuffle is on - play a random song
            Random rand = new Random();
            currentSongIndex = rand.nextInt((mArrSongs.size() - 1) + 1);
            playSong(currentSongIndex);
            //set albumArt and change statusBar color when changed song
            changeMusicAlbumArt(currentSongIndex);
        } else {
            // no repeat or shuffle ON - play next song
            if (currentSongIndex < (mArrSongs.size() - 1)) {
                playSong(currentSongIndex + 1);
                //set albumArt and change statusBar color when changed song
                changeMusicAlbumArt(currentSongIndex + 1);
                currentSongIndex = currentSongIndex + 1;
            } else {
                // play first song
                playSong(0);
                currentSongIndex = 0;
            }
        }
    }

    public void changeMusicAlbumArt(int currentSongIndex) {

        mImgCurrentPlaySong.setImageBitmap(mArrSongs.get(currentSongIndex).getAlbumArt());

        Palette.from(mArrSongs.get(currentSongIndex).getAlbumArt()).generate(new Palette.PaletteAsyncListener() {
            @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
            @Override
            public void onGenerated(Palette palette) {
                Palette.Swatch vibrantSwatch = palette.getVibrantSwatch();
                if (vibrantSwatch != null) {
                    getWindow().setStatusBarColor(vibrantSwatch.getRgb());
                }
            }
        });
    }

    @Override
    public void onBackPressed() {
        if (mSlidingUpLayout != null &&
                (mSlidingUpLayout.getPanelState() == PanelState.EXPANDED || mSlidingUpLayout.getPanelState() == PanelState.ANCHORED)) {
            mSlidingUpLayout.setPanelState(PanelState.COLLAPSED);
        } else {
            super.onBackPressed();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        if (mMediaPlayer != null) {
            if (mMediaPlayer.isPlaying()) {
                mMediaPlayer.stop();
            }
            mMediaPlayer.release();
            mMediaPlayer = null;
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        //Gather permission grant results
        if (requestCode == REQUEST_CODE_PERMISSION) {
            int deniedCount = 0;
            for (int granted : grantResults) {
                //add only permissions which are denied
                if (granted == PackageManager.PERMISSION_DENIED) {
                    deniedCount++;
                }
            }
            if (deniedCount > 0) {
                requestAppPermissions();
            }
        }
    }
}

File: app/src/main/java/com/example/musicplayer/activity/MainActivity.kt
package com.example.musicplayer.activity


import android.annotation.SuppressLint
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.media.MediaPlayer
import android.media.PlaybackParams
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.IBinder
import android.view.View
import android.widget.ImageView
import android.widget.SeekBar
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.musicplayer.R
import com.example.musicplayer.adapter.AdapterSongs
import com.example.musicplayer.constants.PlayerConstants
import com.example.musicplayer.model.Song
import com.example.musicplayer.service.MusicService
import com.example.musicplayer.utils.SpUtility
import com.example.musicplayer.utils.Utilities
import com.sothree.slidinguppanel.SlidingUpPanelLayout.PanelSlideListener
import com.sothree.slidinguppanel.SlidingUpPanelLayout.PanelState
import io.fabric.sdk.android.services.concurrency.AsyncTask
import kotlinx.android.synthetic.main.activity_main.*
import java.io.IOException
import java.util.*


class MainActivity : AppCompatActivity(), View.OnClickListener, MediaPlayer.OnCompletionListener, SeekBar.OnSeekBarChangeListener {

//    private var mArrSongs = ArrayList<Song>()
//    private var mMediaPlayer: MediaPlayer? = null
    private var mHandler: Handler? = null
    private var musicService: MusicService? = null
    private var serviceBound = false

//    private var currentSongIndex = 0

    companion object {
        const val ACTION_PLAY = "com.mahipal.mediaplayer.action.play"
        const val ACTION_PAUSE = "com.mahipal.mediaplayer.action.pause"
        const val ACTION_PREV = "com.mahipal.mediaplayer.action.prev"
        const val ACTION_NEXT = "com.mahipal.mediaplayer.action.next"
        const val ACTION_STOP = "com.mahipal.mediaplayer.action.stop"
    }

    private val mUpdateTimeTask = object : Runnable {
        @SuppressLint("SetTextI18n")
        override fun run() {

            musicService?.mMediaPlayer?.let { player ->

                try {
                    val totalDuration = player.duration.toLong()
                    val currentDuration = player.currentPosition.toLong()

                    // Displaying Total Duration time
                    txt_song_TotalTime?.text = "" + Utilities.milliSecondsToTimer(totalDuration)
                    // Displaying time completed playing
                    txt_song_Curr_Time?.text = "" + Utilities.milliSecondsToTimer(currentDuration)

                    // Updating progress bar
                    val progress = Utilities.getProgressPercentage(currentDuration, totalDuration)
                    //Log.d("Progress", ""+progress);
                    seekBar_play_song?.progress = progress

                    // Running this thread after 100 milliseconds
                    mHandler?.postDelayed(this, 100)
                } catch (ex:IllegalStateException) {
                    ex.printStackTrace()
                }
            }
        }
    }

    private val serviceConnection = object :ServiceConnection {
        override fun onServiceConnected(componentName: ComponentName?, service: IBinder?) {
            val binder = service as MusicService.LocalBinder
            musicService = binder.service
            serviceBound = true

            musicService?.notificationControl = { notifyAction ->
                if (notifyAction.equals(MusicService.NOTIFY_NEXT,true) or
                        notifyAction.equals(MusicService.NOTIFY_PREV,true)) {

                    updateProgressBar()
                }
                updateUiControls()
            }
        }

        override fun onServiceDisconnected(componentName: ComponentName?) {
            musicService = null
            serviceBound = false
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
//        Thread.setDefaultUncaughtExceptionHandler(DefaultExceptionHandler(this))
        setContentView(R.layout.activity_main)

        setSupportActionBar(toolbar)

        AsyncTaskThread().execute()
    }

    private fun updateUi() {
        cv_progress_bar?.visibility = View.GONE

        mHandler = Handler()
//        mMediaPlayer = MediaPlayer()

        setListener()

        sliding_layout?.addPanelSlideListener(object : PanelSlideListener {

            override fun onPanelSlide(panel: View, slideOffset: Float) {

                if (slideOffset > 0.88) {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                        relative?.setBackgroundColor(resources.getColor(R.color.whiteTransparent, theme))
                    }

                    imgBtn_play?.visibility = View.INVISIBLE
                    img_play_song?.scaleType = ImageView.ScaleType.CENTER_CROP
                }
                if (slideOffset < 0.09) {
                    imgBtn_play?.visibility = View.VISIBLE
                    img_play_song?.setImageResource(R.drawable.music)
//                        window.statusBarColor = resources.getColor(R.color.colorPrimaryDark,theme)
                }
            }

            override fun onPanelStateChanged(panel: View, previousState: PanelState, newState: PanelState) {

            }
        })

        recyclerViewSongList?.layoutManager = LinearLayoutManager(
                this, LinearLayoutManager.VERTICAL, false)
        val mAdapterSongs = AdapterSongs(PlayerConstants.SONG_LIST)
        recyclerViewSongList?.adapter = mAdapterSongs

//        currentSongIndex = SpUtility.getInstance(this)?.getCurrenSongIndex() ?: 0

//        Utilities.setImageByByteArray(this, mArrSongs[currentSongIndex.plus(1)].albumArtByteArray, img_playing_song)
//        txt_playing_songName?.text = mArrSongs[currentSongIndex.plus(1)].songTitle
//        imgBtn_play?.setImageResource(R.drawable.ic_play_arrow_24dp)
//        changeMusicAlbumArt(currentSongIndex)
//        btn_song_play?.setImageResource(R.drawable.ic_play_arrow_24dp)

        mAdapterSongs.onItemClick = { adapterPosition ->
            PlayerConstants.SONG_NUMBER = adapterPosition
            PlayerConstants.SONG_PAUSED = false
            playAudio(adapterPosition)
            updateUiControls()
        }
    }

    private fun setListener() {
        try {
            seekBar_play_song?.setOnSeekBarChangeListener(this)
//            mMediaPlayer?.setOnCompletionListener(this)

            imgBtn_play?.setOnClickListener(this)
            btn_song_play?.setOnClickListener(this)
            btn_song_next?.setOnClickListener(this)
            btn_song_previous?.setOnClickListener(this)
            btn_song_shuffle_play?.setOnClickListener(this)
            btn_song_repeat_play?.setOnClickListener(this)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    override fun onClick(v: View) {
        when (v.id) {

            R.id.imgBtn_play -> {
                musicService?.playMedia()
                updateUiControls()
                musicService?.buildNotification()
            }

            R.id.btn_song_play -> {
                musicService?.playMedia()
                updateUiControls()
                musicService?.buildNotification()
            }

            R.id.btn_song_next -> {
                musicService?.nextMedia()
//                updateUiControls()
            }
            R.id.btn_song_previous -> {
                musicService?.prevMedia()
//                updateUiControls()
            }
            R.id.btn_song_shuffle_play ->

                if (PlayerConstants.isShuffle) {
                    PlayerConstants.isShuffle = false
                    Toast.makeText(applicationContext, "Shuffle OFF", Toast.LENGTH_SHORT).show()
                    btn_song_shuffle_play?.setImageResource(R.drawable.ic_shuffle_off_24dp)
                } else {

                    PlayerConstants.isShuffle = true
                    SpUtility.getInstance(this)?.setSongShuffle(PlayerConstants.isShuffle)

                    Toast.makeText(applicationContext, "Shuffle ON", Toast.LENGTH_SHORT).show()

                    PlayerConstants.isRepeat = false
                    SpUtility.getInstance(this)?.setSongRepeat(PlayerConstants.isRepeat)

                    btn_song_shuffle_play?.setImageResource(R.drawable.ic_shuffle_on_24dp)
                    btn_song_repeat_play?.setImageResource(R.drawable.ic_repeat_off_24dp)
                }

            R.id.btn_song_repeat_play ->

                if (PlayerConstants.isRepeat) {

                    PlayerConstants.isRepeat = false
                    SpUtility.getInstance(this)?.setSongRepeat(PlayerConstants.isRepeat)

                    Toast.makeText(applicationContext, "Repeat OFF", Toast.LENGTH_SHORT).show()
                    btn_song_repeat_play?.setImageResource(R.drawable.ic_repeat_off_24dp)
                } else {
                    // make repeat to true
                    PlayerConstants.isRepeat = true
                    SpUtility.getInstance(this)?.setSongRepeat(PlayerConstants.isRepeat)

                    Toast.makeText(applicationContext, "Repeat ON", Toast.LENGTH_SHORT).show()

                    // make shuffle to false
                    PlayerConstants.isShuffle = false
                    SpUtility.getInstance(this)?.setSongShuffle(PlayerConstants.isShuffle)

                    btn_song_repeat_play?.setImageResource(R.drawable.ic_repeat_on_24dp)
                    btn_song_shuffle_play?.setImageResource(R.drawable.ic_shuffle_off_24dp)
                }
        }
    }

    /**
     * Update timer on seekbar
     */
    private fun updateProgressBar() {
        mHandler?.postDelayed(mUpdateTimeTask, 100)
    }

    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {

    }

    override fun onStartTrackingTouch(seekBar: SeekBar) {
        mHandler?.removeCallbacks(mUpdateTimeTask)
    }

    override fun onStopTrackingTouch(seekBar: SeekBar) {

        mHandler?.removeCallbacks(mUpdateTimeTask)

        musicService?.mMediaPlayer?.let { player ->
            val totalDuration = player.duration
            val currentPosition = Utilities.progressToTimer(seekBar.progress, totalDuration)

            // forward or backward to certain seconds
            player.seekTo(currentPosition)

            // update timer progress again
            updateProgressBar()
        }
    }

    override fun onResume() {
        super.onResume()

//        val isServiceRunning = Utilities.isServiceRunning(MusicService::class.java.simpleName,this)
//        if (serviceBound) {
            updateUiControls()
//        }

        updateProgressBar()

    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putBoolean("ServiceState",serviceBound)
        super.onSaveInstanceState(outState)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        serviceBound = savedInstanceState.getBoolean("ServiceState")
    }

    fun updateUiControls() {
        try {

            changeMusicAlbumArt(PlayerConstants.SONG_NUMBER,img_playing_song)
            changeMusicAlbumArt(PlayerConstants.SONG_NUMBER,img_play_song)
            txt_playing_songName?.text = PlayerConstants.SONG_LIST?.get(PlayerConstants.SONG_NUMBER)?.songTitle

        } catch (ex:Exception) {
            ex.printStackTrace()
        }

        if (!PlayerConstants.SONG_PAUSED) {
            imgBtn_play?.setImageResource(R.drawable.ic_pause_24dp)
            btn_song_play?.setImageResource(R.drawable.ic_pause_24dp)
        } else {
            imgBtn_play?.setImageResource(R.drawable.ic_play_arrow_24dp)
            btn_song_play?.setImageResource(R.drawable.ic_play_arrow_24dp)
        }
    }

    override fun onCompletion(mp: MediaPlayer) {
    }

    private fun changeMusicAlbumArt(currentSongIndex: Int,imageView: ImageView) {
        Utilities.setImageByByteArray(this, PlayerConstants.SONG_LIST?.get(PlayerConstants.SONG_NUMBER)?.albumArtByteArray,
                imageView)
    }

    override fun onDestroy() {
        if (PlayerConstants.SONG_PAUSED && musicService != null) {

            mHandler?.removeCallbacks(mUpdateTimeTask)
            musicService?.stopSelf()
            unbindService(serviceConnection)
        }
        super.onDestroy()
    }

    private fun playAudio(songIndex:Int) {
        //Check is service is active
        if (!serviceBound) {

            SpUtility.getInstance(this)?.setCurrentSongIndex(songIndex)

            val playerIntent = Intent(this, MusicService::class.java)
            startService(playerIntent)
            bindService(playerIntent, serviceConnection, Context.BIND_AUTO_CREATE)
        } else {
            //Store the new songIndex to SharedPreferences
            SpUtility.getInstance(this)?.setCurrentSongIndex(songIndex)

            //Service is active
            //Send a broadcast to the service -> Play new Song
            val broadcastIntent = Intent(ACTION_PLAY)
            sendBroadcast(broadcastIntent)
        }

        seekBar_play_song?.progress = 0
        seekBar_play_song?.max = 100
        updateProgressBar()
    }

    override fun onBackPressed() {
        if (sliding_layout != null && (sliding_layout?.panelState == PanelState.EXPANDED ||
                        sliding_layout?.panelState == PanelState.ANCHORED)) {
            sliding_layout?.panelState = PanelState.COLLAPSED
        } else {
            super.onBackPressed()
        }
    }

    inner class AsyncTaskThread : AsyncTask<Unit, Unit, Unit>() {

        override fun doInBackground(vararg p0: Unit?) {
            val mArrSongs = Utilities.getMp3Songs(this@MainActivity)
            PlayerConstants.SONG_LIST = mArrSongs
            SpUtility.getInstance(this@MainActivity)?.storeSongs(mArrSongs)
        }

        override fun onPostExecute(result: Unit?) {
            super.onPostExecute(result)

            updateUi()
        }
    }
}
